akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala

```diff
@@ -255,13 +265,30 @@ private[pool] object SlotState {
 
   }
   final case class WaitingForEndOfResponseEntity(
-    ongoingRequest:  RequestContext,
-    ongoingResponse: HttpResponse) extends ConnectedState with BusyWithResultAlreadyDispatched {
+    ongoingRequest:               RequestContext,
+    ongoingResponse:              HttpResponse,
+    waitingForEndOfRequestEntity: Boolean) extends ConnectedState with BusyWithResultAlreadyDetermined {
 
     override def onResponseEntityCompleted(ctx: SlotContext): SlotState =
-      if (ctx.willCloseAfter(ongoingResponse) || ctx.isConnectionClosed)
+      if (waitingForEndOfRequestEntity)
+        WaitingForEndOfRequestEntity
+      // TODO can we be *sure* that by skipping to Unconnected if ctx.willCloseAfter(ongoingResponse)
+      // we can't get a connection closed event from the 'previous' connection later?
+      else if (ctx.willCloseAfter(ongoingResponse) || ctx.isConnectionClosed)
```

Comment:
not entirely sure but presence of a close header kind of implies that after this request a close is expected... though you may have dug deeper into the state machine here and see something I don't?

Reply:
I agree the presence of a close header implies that after this request a close is expected.

What I'm afraid of is that since this `Slot` will be re-used for the next connection, there might be a race condition in which we'll get the `onConnectionCompleted` of the 'previous' connection while we've already moved on to the 'next' connection.

This is pretty much unrelated to this PR but just an observation I ran into, though.

Reply:
Good point and it certainly plagued me before as well (maybe the legacy pool as well). It is prevented by using `withSlot` inside the `Connection` which first checks if the connection is still the current one of the slot.